subsys/fs/littlefs_fs.c

```diff
@@ -541,154 +541,236 @@ static bool get_page_cb(const struct flash_pages_info *info, void *ctxp)
+{
+	struct fs_littlefs_device_services *device_services = &fs->device_services;
+	struct lfs_config *lcp = &fs->cfg;
+	size_t not_set_function_count = 0;
+	const size_t total_function_count = 8;
 
-	if (prog_size == 0) {
-		prog_size = CONFIG_FS_LITTLEFS_PROG_SIZE;
+	if (!lcp->read) {
+		not_set_function_count++;
 	}
 
-	/* Yes, you can override block size. */
-	lfs_size_t block_size = lcp->block_size;
+	if (!lcp->prog) {
+		not_set_function_count++;
+	}
 
-	if (block_size == 0) {
-		block_size = get_block_size(fs->area);
+	if (!lcp->erase) {
+		not_set_function_count++;
 	}
-	if (block_size == 0) {
-		__ASSERT_NO_MSG(block_size != 0);
-		ret = -EINVAL;
-		goto out;
+
+	if (!lcp->sync) {
+		not_set_function_count++;
 	}
 
-	int32_t block_cycles = lcp->block_cycles;
+	if (!device_services->open) {
+		not_set_function_count++;
+	}
 
-	if (block_cycles == 0) {
-		block_cycles = CONFIG_FS_LITTLEFS_BLOCK_CYCLES;
+	if (!device_services->close) {
+		not_set_function_count++;
 	}
-	if (block_cycles <= 0) {
-		/* Disable leveling (littlefs v2.1+ semantics) */
-		block_cycles = -1;
+
+	if (!device_services->get_block_size) {
+		not_set_function_count++;
 	}
 
-	lfs_size_t cache_size = lcp->cache_size;
+	if (!device_services->get_size) {
+		not_set_function_count++;
+	}
+
+	bool completely_custom = not_set_function_count == 0;
+	bool partly_custom = not_set_function_count != total_function_count &&
+		not_set_function_count != 0;
+
+	LOG_DBG("%zu out of %zu LFS device operations are custom",
+		not_set_function_count, total_function_count);
 
-	if (cache_size == 0) {
-		cache_size = CONFIG_FS_LITTLEFS_CACHE_SIZE;
+	if (completely_custom) {
+		return 0;
 	}
 
-	lfs_size_t lookahead_size = lcp->lookahead_size;
+	if (partly_custom) {
+		LOG_ERR("LFS device operations are only partly custom");
+		return -EINVAL;
+	}
 
-	if (lookahead_size == 0) {
-		lookahead_size = CONFIG_FS_LITTLEFS_LOOKAHEAD_SIZE;
+	lcp->read = fs_littlefs_flash_read;
+	lcp->prog = fs_littlefs_flash_prog;
+	lcp->erase = fs_littlefs_flash_erase;
+	lcp->sync = fs_littlefs_flash_sync;
+	device_services->open = fs_littlefs_flash_open;
+	device_services->close = fs_littlefs_flash_close;
+	device_services->get_block_size = fs_littlefs_flash_get_block_size;
+	device_services->get_size = fs_littlefs_flash_get_size;
+	return 0;
+}
+
+static int littlefs_mount(struct fs_mount_t *mountp)
+{
+	int ret;
+	struct fs_littlefs *fs = mountp->fs_data;
+
+	LOG_INF("LittleFS version %u.%u, disk version %u.%u",
+		LFS_VERSION_MAJOR, LFS_VERSION_MINOR,
+		LFS_DISK_VERSION_MAJOR, LFS_DISK_VERSION_MINOR);
+
+	if (fs->cfg.context) {
+		return -EBUSY;
 	}
 
+	ret = lfs_config_initialize_operations(fs);
+	if (ret < 0) {
+		LOG_ERR("can't initialize configuration (LFS %d)", ret);
+		goto out;
+	}
 
-	/* No, you don't get to override this. */
-	lfs_size_t block_count = fs->area->fa_size / block_size;
+	__ASSERT(fs->cfg.read, "the read function for littlefs is not set");
```

Comment:
Can't these assert's be replaced with a `BUILD_ASSERT` in the GP macro?

Reply:
There is a build assert as well in the macro, but this is basically just a sanity check at the end.

Reply:
Isn't this just taking up ROM space? The error code for a BUILD_ASSERT is not taking any space.

What is the error that should he catched by evaluating this twice ?

Reply:
Programming error, which I consider possible all my shuffling around of function pointers. I do not trust myself, basically. And the assertion macros which I have used so far in my life are usually made in a way, that they can be optimized out for a release build. Is this not the case for __ASSERT?

Reply:
Maybe the are optimized out for a release build. There is no reason to check it twice and maybe a third time before calling the routine (I didn't check if this is the case). These checks IMO tend to make the code less readable.

Reply:
It will not defend you from "shuffling" erase into "fs->cfg.prog".

Reply:
Convincing arguments, it is done and pushed